classes/village/VillageApiPresenter.php

```diff
@@ -255,4 +200,104 @@ function ($row) {
+                    "construction_time" => $villageBuildingDto->construction_time,
+                    "construction_time_remaining" => $villageBuildingDto->construction_time_remaining,
+                    "requirements_met" => $villageBuildingDto->requirements_met,
+                    "upgrade_sets" => array_map(
+                        function ($villageUpgradeSetDto) {
+                            return [
+                                "key" => $villageUpgradeSetDto->key,
+                                "name" => $villageUpgradeSetDto->name,
+                                "description" => $villageUpgradeSetDto->description,
+                                "upgrades" => array_map(
+                                    function ($villageUpgradeDto) {
+                                        return [
+                                            "id" => $villageUpgradeDto->id,
+                                            "key" => $villageUpgradeDto->key,
+                                            "village_id" => $villageUpgradeDto->village_id,
+                                            "status" => $villageUpgradeDto->status,
+                                            "research_progress" => $villageUpgradeDto->research_progress,
+                                            "research_progress_required" => $villageUpgradeDto->research_progress_required,
+                                            "research_progress_last_updated" => $villageUpgradeDto->research_progress_last_updated,
+                                            "research_boosted" => $villageUpgradeDto->research_boosted,
+                                            "name" => $villageUpgradeDto->name,
+                                            "tier" => $villageUpgradeDto->tier,
+                                            "description" => $villageUpgradeDto->description,
+                                            "materials_research_cost" => $villageUpgradeDto->materials_research_cost,
+                                            "food_research_cost" => $villageUpgradeDto->food_research_cost,
+                                            "wealth_research_cost" => $villageUpgradeDto->wealth_research_cost,
+                                            "research_time" => $villageUpgradeDto->research_time,
+                                            "research_time_remaining" => $villageUpgradeDto->research_time_remaining,
+                                            "food_upkeep" => $villageUpgradeDto->food_upkeep,
+                                            "materials_upkeep" => $villageUpgradeDto->materials_upkeep,
+                                            "wealth_upkeep" => $villageUpgradeDto->wealth_upkeep,
+                                            "requirements_met" => $villageUpgradeDto->requirements_met,
+                                        ];
+                                    },
```

Comment:
This looks like a good candidate to be its own presenter and keep this function from being too large